Database/MySQL/PreparedStatement.hs

```diff
@@ -0,0 +1,485 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+module Database.MySQL.PreparedStatement
+    where
+
+import           Control.Monad             (void, when)
+import           Data.ByteString.Char8
+import           Data.ByteString.Internal  (create)
+import           Data.ByteString.Unsafe    (unsafeUseAsCStringLen)
+import qualified Data.Fixed                as Fixed
+import           Data.Text                 (Text)
+import qualified Data.Text.Foreign         as Text
+import           Data.Time.Calendar        (Day, fromGregorian, toGregorian)
+import           Data.Time.LocalTime       (LocalTime (..), TimeOfDay (..),
+                                            makeTimeOfDayValid)
+import           Database.MySQL.Base.C
+import           Database.MySQL.Base.Types hiding (Type (..))
+import qualified Database.MySQL.Base.Types as Types
+import           Foreign                   hiding (newForeignPtr, void)
+import           Foreign.C.Error
+import           Foreign.C.String          (peekCString)
+import           Foreign.Concurrent        (newForeignPtr)
+
+data Statement = Statement
+    { connectionPtr :: ForeignPtr MYSQL
+    , statementPtr  :: ForeignPtr MYSQL_STMT
+    }
+
+prepare :: ForeignPtr MYSQL -> Text -> IO Statement
+prepare mysql q =
+    withForeignPtr mysql $ \mysql' -> do
+    mysqlStmt <- throwIfNull "mysql_stmt_init" $ mysql_stmt_init mysql'
+    stmt <- newForeignPtr mysqlStmt (void $ mysql_stmt_close mysqlStmt)
+    Text.withCStringLen q $ \(p,l) -> do
+        res <- mysql_stmt_prepare mysqlStmt p (fromIntegral l)
+        when (res > 0) $ do
+            err <- peekCString =<< mysql_stmt_error mysqlStmt
+            error err
+    pure $ Statement mysql stmt
+
+withConn :: Statement -> (Ptr MYSQL -> IO a) -> IO a
```

Comment:
the pointer to the underlying MYSQL is only needed for a `mysql_error` call which is only relevant for `mysql_stmt_close` all other functions will call `mysql_stmt_error`

Should we stop holding on to the ref for the duration of the Statement? Is it possible for the Connection to get dropped if `Statement` doesn't hold it?